Remove three ObjC metadata fields from non-ObjC runtime

davezarzycki opened the pull request:
Hi @rjmccall – Before I finish this draft pull request, I'd like to verify that:

1) The non-Apple ABI is still not locked down.
2) This change is reasonable.

I'll finish fixing all the broken tests if this is the case. Thanks! As a bonus, what other fields are good candidates for removal from the non-ObjC runtime?

**[EDIT]** – A related followup question for a followup PR: Should the vtable layout be more like C++? I.e. where the positive displacements are vtable slots and the negative displacements are language/runtime metadata? This would allow for better code density.

davezarzycki committed:
Remove three ObjC fields from non-ObjC runtime

mikeash commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -2623,11 +2623,15 @@ class MetadataReader {
                              &dataPtr))
       return StoredPointer();
+#else
+    StoredPointer dataPtr = SWIFT_CLASS_IS_SWIFT_MASK;
```
You can just `return StoredPointer();` here. You'll end up hitting the `if (!dataPtr)` path a few lines later anyway.

mikeash reviewed (approved):
Seems reasonable to me.

rjmccall commented:
Ideally, MetadataReader wouldn't be sensitive to target #if conditions, and instead ObjC interop-ness would be reflected in the Target abstraction.  That would e.g. allow tools built one way to debug/analyze programs built the other.  To make that work, you'll have to play some template games in Metadata.h, but it shouldn't be too bad.

jckarter commented:
You are correct that the non-Darwin ABI is still up for grabs. I agree with @rjmccall that it'd be nice if we could make this dependent on the `Runtime` template argument instead of using conditional compilation.

A good followup for non-Darwin optimization would be to start treating class objects as being directly referenceable as metadata, like struct/enum metadata is, since there's no ObjC realization to worry about. That should save a good amount of code size from metadata accesses, and allow dynamic class metadata instantiation for generic or resilient classes to be more efficient and avoid some synchronization that's necessary on Darwin to coordinate with ObjC realization.

rjmccall commented:
Yeah.  It'll need some new logic in order to check for transitive direct-referenceability of superclasses, since right now the condition is just "no, you can't emit this reference directly", but it should be possible to be much more efficient than we are today.

compnerd commented on lib/IRGen/ClassMetadataVisitor.h:
```diff
@@ -65,8 +65,10 @@ template <class Impl> class ClassMetadataVisitor
     // FIXME: Figure out what can be removed altogether in non-objc-interop
     // mode and remove it. rdar://problem/18801263
```
Does this not address this FIXME?

davezarzycki replied:
I'm not sure and that's why I asked at the top: "what other fields are good candidates for removal from the non-ObjC runtime?" 

I think that @rjmccall and @jckarter got too excited about other things that need fixing, so they missed that question. ;-)

jckarter replied:
The three extra padding fields seem like the biggest obvious thing to remove. I can't off the top of my head think of any other places where we could wholesale remove fields. I'd be fine removing the fixme.

davezarzycki commented:
Hi @rjmccall and @jckarter – I'd like to do incremental updates. Would you be okay with focusing in unnecessary field removal in this PR? In particular, the MetadataReader was already using ifdefs, so I'd like to leave fixing that issue to a later PR.

rjmccall commented:
> I'm not sure and that's why I asked at the top: "what other fields are good candidates for removal from the non-ObjC runtime?"

That addresses it for classes, which is the scope of this particular FIXME.

Assuming that we're not interested in highly-disruptive layout changes like putting the VWT at a positive offset, I think most of the other ways that ObjC inserts itself into the runtime representation are just extra alternatives in various metadata references — things like protocol references possibly being to ObjC protocols, or type references possibly being to ObjC classes.

> Would you be okay with focusing in unnecessary field removal in this PR? In particular, the MetadataReader was already using ifdefs, so I'd like to leave fixing that issue to a later PR.

I think it's fine to make MetadataReader work cross-ObjC-interop in a different patch.  Please base the conditionalization in Metadata.h on the Runtime parameter, though, otherwise it's just not being done right.

jckarter commented:
If we base the presence of the ObjC compatibility fields on the `Runtime` template parameter, then the correct MetadataReader behavior ought to fall out automatically.

rjmccall commented:
For this, at least.  Dave is right that there are existing `#if`s in that file which ought to be converted to uses of information from `Runtime` if we're going to support running cross-configuration properly.

davezarzycki commented:
@swift-ci please test

davezarzycki commented:
@swift-ci please test windows

davezarzycki commented:
> A good followup for non-Darwin optimization would be to start treating class objects as being directly referenceable as metadata, like struct/enum metadata is, since there's no ObjC realization to worry about.

Hi @jckarter – If you point me at a few lines of code to see the difference you're talking about, then I can make an attempt at fixing it.

davezarzycki commented:
@swift-ci please test macos

davezarzycki commented:
Hi @compnerd – Does `FileCheck` not execute `RUN` directives via a POSIX compliant shell on Windows? I'm trying to use `$(( 1 + 2 * 3 ))` style shell arithmetic in a `RUN` line.

davezarzycki commented:
@swift-ci please test

compnerd commented:
Right, there is no posix shell on windows, the tests on windows are rub in lits shell emulation.

davezarzycki commented:
Do you know if the % escape/substitution mechanism in the test suite allows arithmetic?

davezarzycki commented:
@swift-ci please test
@swift-ci please test windows

davezarzycki commented:
@swift-ci please test windows

davezarzycki commented:
@swift-ci please clean test windows

davezarzycki commented:
@swift-ci please test windows

rjmccall commented:
I think Joe came up with a way to do some preprocessing on check files that might work as an alternative to whatever you're trying to do.

davezarzycki commented:
@swift-ci please test

davezarzycki commented:
Hi @compnerd – I've changed some of the tests in this PR to explicitly test with and without ObjC interoperability whereas before they just tested whatever %target-runtime was set to. This causing failures on Windows. What should I do?

davezarzycki commented:
@swift-ci please clean test macos

davezarzycki commented:
@swift-ci please test

davezarzycki commented:
@swift-ci please test windows

davezarzycki commented:
@swift-ci please test windows

davezarzycki commented:
@swift-ci please test

jckarter commented:
@davezarzycki If you want to look into making class metadata statically addressable, a good place to start looking would be `shouldCacheTypeMetadataAccess` and `shouldTypeMetadataAccessUseAccessor` in IRGen. You could also look over the `*ClassMetadataBuilder*` classes in IRGen and the various `swift_init*Class*` functions in the runtime to see if there are opportunities for optimizing the initialization process for generic or resilient classes when ObjC realization is not necessary.

compnerd commented:
@swift-ci please test Windows platform

compnerd commented:
@swift-ci please test Windows platform

compnerd committed:
test: account for indirect imports

Adjust the test to accommodate environments which do indirect imports,
which use the singleton strategy, relying on the runtime initialization
of fields which are imported.

compnerd commented:
#31981 has the test adjustments needed for this.  Could you please grab the top commit and put it into this PR?

davezarzycki commented:
@swift-ci please test Windows platform

davezarzycki commented:
@swift-ci please test

davezarzycki commented:
@swift-ci please test Windows platform

davezarzycki commented:
@swift-ci please test Windows platform

davezarzycki commented:
@compnerd The windows failure looks unrelated. Agreed? Ready to merge?

@swift-ci please test Windows platform

compnerd commented:
Yeap, that looks unrelated to this change.  _sigh_ time to go track that down.
